### PR TITLE
Add DynamicCaseMapping to DynamicOptions

### PR DESCRIPTION
**Pull request by annelo-msft**
## Context

In the [recent arch board discussion of DynamicData](https://github.com/Azure/azure-sdk-for-net/issues/35944), suggestions were made to:
- [ ] Rework DynamicDataOptions.Default and DynamicDataNameMapping enum such that
  - [ ] anything called Default has the same behavior as not passing any parameteters
  - [ ] a good default experience works for both read/write of the JSON
  - [ ] enum names make it clear what the resulting behavior of the type will be

A decision was made to remove the DynamicDataNameMapping entirely and implemented in https://github.com/Azure/azure-sdk-for-net/issues/35944.  Concerns were raised about potential data loss, and this decision was overturned.  Further discussion can be found [here](https://gist.github.com/annelo-msft/6ce196664d47bdda7e696144123a84e2).

## This PR:
- Implements support for strict pass-through naming as the default DynamicData experience
- Adds a DynamicCaseMapping enum value to DynamicDataOptions to allow callers to specify they would like to use the dynamic content value as an idiomatic .NET type
- Updates the ToCamelCase logic to handle more cases, including some raised as specific concerns
- Restructures case-mapping tests and moves all non-casing tests to idiomatic JSON style conventions

Addresses https://github.com/Azure/azure-sdk-for-net/issues/35944

### CODE REVIEW

File: sdk/core/Azure.Core.Experimental/tests/DynamicData/DynamicCaseMappingTests.cs [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+            Assert.AreEqual(true, (bool)value.ParentCamel.NestedPascal);
+            Assert.AreEqual("c", (string)value.ParentPascal.NestedPascal);
+
+            Assert.AreEqual("new", (string)value["Pascal"]);
+            Assert.AreEqual(true, (bool)value.ParentCamel["NestedPascal"]);
+            Assert.AreEqual("c", (string)value.ParentPascal["NestedPascal"]);
+        }
+
+        [Test]
+        public void CamelCaseMappingWorksForConcerningCases()
```
Comment (annelo-msft) [inline sdk/core/Azure.Core.Experimental/tests/DynamicData/DynamicCaseMappingTests.cs:402]: This one is for you, @heaths!

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/tests/ConversationAnalysisClientLiveTests.cs [hunk 2]
```diff
@@ -45,7 +46,8 @@ public async Task AnalyzeConversation()
             Assert.IsNotNull(response);
 
             // deserialize
-            dynamic conversationalTaskResult = response.Content.ToDynamicFromJson();
+            DynamicDataOptions options = new() { CaseMapping = DynamicCaseMapping.PascalToCamel };
```
Comment (heaths) [inline sdk/cognitivelanguage/Azure.AI.Language.Conversations/tests/ConversationAnalysisClientLiveTests.cs:49]: @annelo-msft would you like me to change the tests (the samples should be fine already, IIRC) to match our defaults in case customers look at them? I could do that in a separate PR so it's not a bother for you if so.

File: sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicCaseMapping.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    /// </summary>
+    public enum DynamicCaseMapping
+    {
+        /// <summary>
+        /// Properties are read from and written to the data buffer with the same casing as the DynamicData property.
+        /// </summary>
+        None = 0,
+
+        /// <summary>
+        /// A "PascalCase" DynamicData property can be used to read and set "camelCase" properties that exist in the data buffer.
```
Comment (heaths) [inline sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicCaseMapping.cs:17]: Thank you for casing the names as such for those less familiar with the casing conventions! 🙏

File: sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicData.cs [hunk 4]
```diff
@@ -44,14 +44,23 @@ internal static JsonSerializerOptions GetSerializerOptions(DynamicDataOptions op
         {
             JsonSerializerOptions serializer = new()
```
Comment (heaths) [inline sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicData.cs:45]: Nit: only because this threw me a little in lines 56 and 57 below, but could we call this `serializationOptions` or just `options`? I knew the properties you were setting as belonging to `JsonSerializerOptions`, but when I saw them on `serializer` I had to look if that was some abstraction we had.

File: sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicData.cs [hunk 5]
```diff
@@ -44,14 +44,23 @@ internal static JsonSerializerOptions GetSerializerOptions(DynamicDataOptions op
...
                 Converters =
                 {
                     new DefaultTimeSpanConverter()
                 }
             };
 
+            switch (options.CaseMapping)
+            {
+                case DynamicCaseMapping.PascalToCamel:
+                    serializer.PropertyNameCaseInsensitive = true;
```
Comment (heaths) [inline sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicData.cs:56]: I don't disagree with this particular line - only want to ask (and if I did on the previous similar PR, I apologize if we already closed on this) if we shouldn't be explicit if they are. That said, since this only affects *de*serialization I think it's fine. Just asking the question. :)

**Review by heaths — APPROVED** [review]

File: sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicCaseMapping.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Azure.Core.Dynamic
+{
+    /// <summary>
+    /// Options for getting and setting DynamicData properties.
```
Comment (christothes) [inline sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicCaseMapping.cs:7]: If we are planning to keep indexer property access as direct pass-through no matter which option is set here, we should document that explicitly here and on the options type.

File: sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicDataOptions.cs [hunk 7]
```diff
@@ -14,6 +14,11 @@ public class DynamicDataOptions
         };
         internal static DynamicDataOptions Default { get => _default; }
 
+        /// <summary>
+        /// Gets or sets an object that specifies how dynamic property names will be mapped to member names in the data buffer.
```
Comment (christothes) [inline sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicDataOptions.cs:18]: nit: "in the data buffer" sounds too implementation detail-y to me. I wonder if we should change terminology slightly to something like "of the underlying data model".

**Review by christothes — APPROVED** [review]

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/tests/ConversationAnalysisClientLiveTests.cs (hunk 2 above)
Comment (annelo-msft) [reply sdk/cognitivelanguage/Azure.AI.Language.Conversations/tests/ConversationAnalysisClientLiveTests.cs:49]: I had understood that this is what we want to do in customer-facing documentation.  What is it that makes you think this should change?

File: sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicData.cs (hunk 4 above)
Comment (annelo-msft) [reply sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicData.cs:45]: Sure, yeah, I just wanted something short and options was already taken.  I can change it.

File: sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicData.cs (hunk 5 above)
Comment (annelo-msft) [reply sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicData.cs:56]: Ah, I see, I think I understand now.  Since we're setting `serializer.PropertyNamingPolicy = JsonNamingPolicy.CamelCase;`, we have a full, more precise specification of the mapping for both serialization and deserialization, and don't need this property to be set?  It looks like that works, so happy to remove this.

File: sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicCaseMapping.cs (hunk 6 above)
Comment (annelo-msft) [reply sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicCaseMapping.cs:7]: Cool, I'll update that.

File: sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicDataOptions.cs (hunk 7 above)
Comment (annelo-msft) [reply sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicDataOptions.cs:18]: I feel like using the word model is tied to a subset of use cases as well, but I'm happy to rework this to remove the word buffer.

File: sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicCaseMapping.cs (hunk 3 above)
Comment (annelo-msft) [reply sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicCaseMapping.cs:17]: 👍

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/tests/ConversationAnalysisClientLiveTests.cs (hunk 2 above)
Comment (annelo-msft) [reply sdk/cognitivelanguage/Azure.AI.Language.Conversations/tests/ConversationAnalysisClientLiveTests.cs:49]: Cool.  As I understand the vision for this type, we want it -- and our docs -- to look as much like idiomatic .NET code as possible.  Given that, I feel it's regrettable that the identified risks resulted in needing to pass options into the type in every instance in order to use it in the intended way -- I wish we had been able to find a solution that improved the customer experience here.  But I also know there are diverging perspectives on how customers will want to use the type, and we have made it possible for them to make an informed decision based on their own preferences.

Comment (annelo-msft) [reply sdk/cognitivelanguage/Azure.AI.Language.Conversations/tests/ConversationAnalysisClientLiveTests.cs]: Updated to simplify a bit.
